src/bip47/Bip47Util.cpp

```diff
@@ -0,0 +1,177 @@
+#include "Bip47Util.h"
+#include "wallet/wallet.h"
+#include "PaymentCode.h"
+#include "SecretPoint.h"
+#include "primitives/transaction.h"
+#include "PaymentAddress.h"
+#include <vector>
+#include "uint256.h"
+
+using namespace std;
+
+bool BIP47Util::getOpCodeOutput(const CTransaction& tx, CTxOut& txout) {
+    for(int i = 0; i < tx.vout.size(); i++) {
+        if (tx.vout[i].scriptPubKey[0] == OP_RETURN) {
+            txout = tx.vout[i];
+            return true;
+        }
+    }
+    return false;
+}
+
+
+bool BIP47Util::isValidNotificationTransactionOpReturn(CTxOut txout) {
+    vector<unsigned char> op_date;
+    return getOpCodeData(txout, op_date);
+}
+
+bool BIP47Util::getOpCodeData(CTxOut txout, vector<unsigned char>& op_data) {
+    CScript::const_iterator pc = txout.scriptPubKey.begin();
+    vector<unsigned char> data;
+    
+    while (pc < txout.scriptPubKey.end())
+    {
+        opcodetype opcode;
+        if (!txout.scriptPubKey.GetOp(pc, opcode, data))
+        {
+            LogPrintf("GetOp false in getOpCodeData\n");
+            return false;
+        }
+        LogPrintf("Data.size() = %d,  opcode = 0x%x\n", data.size(), opcode);
+        if (data.size() > 0 && opcode < OP_PUSHDATA4  )
+        {
+            op_data = data;
+            return true;
+        } 
+        
+    }
+}
+
+bool BIP47Util::getPaymentCodeInNotificationTransaction(vector<unsigned char> privKeyBytes, CTransaction tx, PaymentCode &paymentCode) {
+    // tx.vin[0].scriptSig
+    CWalletTx wtx(pwalletMain, tx);
```

Comment:
I think no need to create additional object here, you can get all data you need below from `CTransaction tx`